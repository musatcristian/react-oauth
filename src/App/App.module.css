main {
  display: flex;
  flex-direction: column;
  margin: 30px;
  padding: 30px;
  min-height: calc(100vh - 120px);
}

.details {
  flex-grow: 1;
  border-radius: 5px;
  box-shadow: 0 3px 10px rgb(70 130 180 / 0.2);
  margin-top: 15px;
  padding: 5px;
  position: relative;
}

.container {
  align-items: flex-start;
  display: flex;
  justify-content: space-between;
  flex: 1;
}
