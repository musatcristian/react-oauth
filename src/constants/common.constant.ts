export const HEADING = 'Newest post from your favorite influencer!';
export const ERROR = "We couldn't find the requested resource";
export const LOGIN = 'Login';
export const LOGIN_GITHUB = 'Login with Github';
