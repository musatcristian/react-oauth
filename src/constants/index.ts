export * from './search.constant';
export * from './common.constant';
