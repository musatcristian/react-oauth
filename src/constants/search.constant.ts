export const SEARCH_HANDLE = 'Search Handle';
export const SEARCH_POST = 'Search Post';
export const SEARCH_LATEST = 'Get latest account info';
export const PLACEHOLDER_POST = 'Search by post shortcode...';
export const PLACEHOLDER_ACCOUNT = 'Search by instagram handle...';
