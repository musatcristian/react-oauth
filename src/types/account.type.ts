export interface IAccountDetails {
  biography: string;
  last_retrieved: Date;
  full_name: string;
  followers: number;
  recent_post: {
    shortcode: string;
    display_url: string;
    likes: number;
    comments: number;
    type: 'image' | 'video' | 'carousel';
  };
}

export interface IAccountDetailsProps {
  details: IAccountDetails | null;
}
