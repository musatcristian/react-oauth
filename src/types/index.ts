export * from './account.type';
export * from './search.type';
export * from './post.type';
export * from './login.type';
