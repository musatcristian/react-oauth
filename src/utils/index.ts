export * from './getAccountDetails.util';
export * from './getPostDetails.util';
