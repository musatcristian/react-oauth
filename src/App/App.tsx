import React, { useCallback, useState } from 'react';

import { IAccountDetails, IPostDetails } from '../types';
import { getAccountDetails, getPostDetails } from '../utils';
import { Search, AccountDetails, PostDetails, NotFound, Loading } from '../components';

export const App: React.FunctionComponent = () => {
  const [accountDetails, setAcountDetails] = useState<IAccountDetails | null>(null);
  const [postDetails, setPostDetails] = useState<IPostDetails | null>(null);
  const [showNotFound, setShowNotFound] = useState<boolean>(false);
  const [loading, setLoading] = useState<boolean>(false);

  const handleAccountSearch = useCallback(async (handle: string, force: boolean) => {
    setShowNotFound(false);
    setLoading(true);
    try {
      const details = await getAccountDetails(handle, force);
      setAcountDetails(details);
    } catch (_error) {
      setShowNotFound(true);
    } finally {
      setLoading(false);
    }
  }, []);

  const handlePostSearch = useCallback(async (shortcode: string) => {
    setShowNotFound(false);
    setLoading(true);
    try {
      const post = await getPostDetails(shortcode);
      setPostDetails(post);
    } catch (error) {
      setShowNotFound(true);
    } finally {
      setLoading(false);
    }
  }, []);

  return (
    <main>
      <div>Newest post from your favorite influencer!</div>
      <Search onAccountSearch={handleAccountSearch} onPostSearch={handlePostSearch} />
      <section>
        {loading && <Loading />}
        {showNotFound && <NotFound />}
        <div>
          <AccountDetails details={accountDetails} />
          <PostDetails details={postDetails} />
        </div>
      </section>
    </main>
  );
};
