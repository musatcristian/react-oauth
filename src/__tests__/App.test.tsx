import React from 'react';
import { findByText, fireEvent, render, screen, waitFor } from '@testing-library/react';
import '@testing-library/jest-dom';

import { App } from '../App';
import { getAccountDetails, getPostDetails, showGithubUser } from '../utils';
import { GithubUser, IAccountDetails, IPostDetails } from '../types';
import { HEADING, LOGIN, LOGIN_GITHUB, PLACEHOLDER_ACCOUNT, PLACEHOLDER_POST, SEARCH_HANDLE, SEARCH_POST } from '../constants';

const mock_account: IAccountDetails = {
  biography: 'biography',
  last_retrieved: new Date(),
  full_name: 'Full Name',
  followers: 2,
  recent_post: {
    shortcode: 'shortcode',
    display_url: 'url',
    likes: 5,
    comments: 3,
    type: 'image',
  },
};

const mock_post: IPostDetails = {
  alt: 'alt definition',
  caption: 'caption',
  dimensions: {
    height: 1000,
    width: 1020,
  },
  display_url: 'url',
};

const mock_user: GithubUser = {
  avatar_url: '',
  bio: 'bio',
  blog: 'asdasd',
  company: 'ggdgsd',
  created_at: 'yesterday',
  email: '@email',
  events_url: 'events',
  followers: 2,
  followers_url: 'gdgdgd',
  following: 2,
  following_url: 'dhdhshds',
  gists_url: 'gist',
  gravatar_id: '123123123',
  hireable: false,
  html_url: 'tatata.url.com',
  id: 1,
  location: 'bababa',
  login: 'login',
  name: 'Cristian Musat',
  node_id: 'hsdhdshhsd',
  organizations_url: 'jdjdjd',
  public_gists: 0,
  public_repos: 5,
  received_events_url: 'hdhdhdhd',
  repos_url: 'repos',
  site_admin: true,
  starred_url: 'jsjsjsjs',
  subscriptions_url: 'asubs',
  twitter_username: 'twitter',
  type: 'user',
  updated_at: 'today',
  url: 'jajjasjajs',
}

jest.mock('../utils/getAccountDetails.util.ts');
jest.mock('../utils/getPostDetails.util.ts');
jest.mock('../utils/github.util.ts');

const mockGetAccount = getAccountDetails as jest.MockedFunction<typeof getAccountDetails>;
const mockGetPost = getPostDetails as jest.MockedFunction<typeof getPostDetails>;
const mockGithubUser = showGithubUser as jest.MockedFunction<typeof showGithubUser>;

describe('<App />', () => {
  it('renders App component', () => {
    const {queryByText} = render(<App />);

    const main = queryByText(HEADING);

    expect(main).toBeVisible();
  });

  it('should show login modal', () => {
    const { getByText } = render(<App />);
    const button = getByText(LOGIN);

    expect(button).toBeVisible();

    fireEvent.click(button);
    const modal = getByText(LOGIN_GITHUB);
    expect(modal).toBeVisible();
  });

  it('should show user name', async () => {
    mockGithubUser.mockResolvedValueOnce(mock_user);
    const { getByText } = render(<App />);
    const showUser = getByText('Show User');
    expect(showUser).toBeVisible();
    fireEvent.click(showUser);

    expect(mockGithubUser).toHaveBeenCalledTimes(1);
    expect(mockGithubUser).toHaveBeenCalledWith();

    await waitFor(() => {
      const user = getByText('Hi ' + mock_user.name);
      expect(user).toBeVisible();
      expect(showUser).not.toBeVisible();
    });
  });

  test('App snapshot', () => {
    const { baseElement } = render(<App />);
    expect(baseElement).toMatchSnapshot();
  });
});
