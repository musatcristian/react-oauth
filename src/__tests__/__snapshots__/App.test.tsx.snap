// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<App /> App snapshot 1`] = `
<body>
  <div>
    <main>
      <h3>
        Newest post from your favorite influencer!
      </h3>
      <button>
        Login
      </button>
      <section>
        <button>
          Show User
        </button>
      </section>
    </main>
  </div>
</body>
`;
